File: src/Heq/Dsp/Biquad.cs
using System;

namespace Heq.Dsp
{
    public struct Biquad
    {
        public double B0, B1, B2, A0, A1, A2;

        public double MagnitudeSquared(double w)
        {
            double s = Math.Sin(w * 0.5);
            double phi = s * s;

            double bSum = B0 + B1 + B2;
            double aSum = A0 + A1 + A2;

            double num = bSum * bSum
                         - 4.0 * (B0 * B1 + 4.0 * B0 * B2 + B1 * B2) * phi
                         + 16.0 * B0 * B2 * phi * phi;
            double den = aSum * aSum
                         - 4.0 * (A0 * A1 + 4.0 * A0 * A2 + A1 * A2) * phi
                         + 16.0 * A0 * A2 * phi * phi;

            if (den <= 0) return 0;
            if (num < 0) num = 0;
            return num / den;
        }

        public double GainDb(double w)
        {
            double m2 = MagnitudeSquared(w);
            return m2 <= 1e-20 ? -200.0 : 10.0 * Math.Log10(m2);
        }

        public static Biquad Peaking(double freq, double sampleRate, double gainDb, double q)
        {
            double w0 = Omega(freq, sampleRate);
            double a = Math.Pow(10.0, gainDb / 40.0);
            double alpha = Math.Sin(w0) / (2.0 * q);
            double cos = Math.Cos(w0);

            return new Biquad
            {
                B0 = 1 + alpha * a,
                B1 = -2 * cos,
                B2 = 1 - alpha * a,
                A0 = 1 + alpha / a,
                A1 = -2 * cos,
                A2 = 1 - alpha / a,
            };
        }

        public static Biquad LowShelf(double freq, double sampleRate, double gainDb, double q)
        {
            double w0 = Omega(freq, sampleRate);
            double a = Math.Pow(10.0, gainDb / 40.0);
            double alpha = Math.Sin(w0) / (2.0 * q);
            double cos = Math.Cos(w0);
            double sqrtA2Alpha = 2.0 * Math.Sqrt(a) * alpha;

            return new Biquad
            {
                B0 = a * ((a + 1) - (a - 1) * cos + sqrtA2Alpha),
                B1 = 2 * a * ((a - 1) - (a + 1) * cos),
                B2 = a * ((a + 1) - (a - 1) * cos - sqrtA2Alpha),
                A0 = (a + 1) + (a - 1) * cos + sqrtA2Alpha,
                A1 = -2 * ((a - 1) + (a + 1) * cos),
                A2 = (a + 1) + (a - 1) * cos - sqrtA2Alpha,
            };
        }

        public static Biquad HighShelf(double freq, double sampleRate, double gainDb, double q)
        {
            double w0 = Omega(freq, sampleRate);
            double a = Math.Pow(10.0, gainDb / 40.0);
            double alpha = Math.Sin(w0) / (2.0 * q);
            double cos = Math.Cos(w0);
            double sqrtA2Alpha = 2.0 * Math.Sqrt(a) * alpha;

            return new Biquad
            {
                B0 = a * ((a + 1) + (a - 1) * cos + sqrtA2Alpha),
                B1 = -2 * a * ((a - 1) + (a + 1) * cos),
                B2 = a * ((a + 1) + (a - 1) * cos - sqrtA2Alpha),
                A0 = (a + 1) - (a - 1) * cos + sqrtA2Alpha,
                A1 = 2 * ((a - 1) - (a + 1) * cos),
                A2 = (a + 1) - (a - 1) * cos - sqrtA2Alpha,
            };
        }

        public static Biquad HighPass(double freq, double sampleRate, double q)
        {
            double w0 = Omega(freq, sampleRate);
            double alpha = Math.Sin(w0) / (2.0 * q);
            double cos = Math.Cos(w0);

            return new Biquad
            {
                B0 = (1 + cos) / 2,
                B1 = -(1 + cos),
                B2 = (1 + cos) / 2,
                A0 = 1 + alpha,
                A1 = -2 * cos,
                A2 = 1 - alpha,
            };
        }

        public static Biquad LowPass(double freq, double sampleRate, double q)
        {
            double w0 = Omega(freq, sampleRate);
            double alpha = Math.Sin(w0) / (2.0 * q);
            double cos = Math.Cos(w0);

            return new Biquad
            {
                B0 = (1 - cos) / 2,
                B1 = 1 - cos,
                B2 = (1 - cos) / 2,
                A0 = 1 + alpha,
                A1 = -2 * cos,
                A2 = 1 - alpha,
            };
        }

        public static Biquad Notch(double freq, double sampleRate, double q)
        {
            double w0 = Omega(freq, sampleRate);
            double alpha = Math.Sin(w0) / (2.0 * q);
            double cos = Math.Cos(w0);

            return new Biquad
            {
                B0 = 1,
                B1 = -2 * cos,
                B2 = 1,
                A0 = 1 + alpha,
                A1 = -2 * cos,
                A2 = 1 - alpha,
            };
        }

        public static Biquad BandPass(double freq, double sampleRate, double q)
        {
            double w0 = Omega(freq, sampleRate);
            double alpha = Math.Sin(w0) / (2.0 * q);
            double cos = Math.Cos(w0);

            return new Biquad
            {
                B0 = alpha,
                B1 = 0,
                B2 = -alpha,
                A0 = 1 + alpha,
                A1 = -2 * cos,
                A2 = 1 - alpha,
            };
        }

        public static Biquad AllPass(double freq, double sampleRate, double q)
        {
            double w0 = Omega(freq, sampleRate);
            double alpha = Math.Sin(w0) / (2.0 * q);
            double cos = Math.Cos(w0);

            return new Biquad
            {
                B0 = 1 - alpha,
                B1 = -2 * cos,
                B2 = 1 + alpha,
                A0 = 1 + alpha,
                A1 = -2 * cos,
                A2 = 1 - alpha,
            };
        }

        private static double Omega(double freq, double sampleRate)
        {
            double f = Math.Max(1.0, Math.Min(freq, sampleRate * 0.4999));
            return 2.0 * Math.PI * f / sampleRate;
        }

        public static double[] ButterworthQs(int order)
        {
            if (order < 2) order = 2;
            if (order % 2 != 0) order++;
            int sections = order / 2;
            var qs = new double[sections];
            for (int k = 0; k < sections; k++)
                qs[k] = 1.0 / (2.0 * Math.Cos((2.0 * k + 1.0) * Math.PI / (2.0 * order)));
            return qs;
        }
    }
}

File: rust/heq-core/src/dsp/biquad.rs
use std::f64::consts::PI;

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct Biquad {
    pub b0: f64,
    pub b1: f64,
    pub b2: f64,
    pub a0: f64,
    pub a1: f64,
    pub a2: f64,
}

impl Biquad {
    pub fn magnitude_squared(&self, w: f64) -> f64 {
        let s = (w * 0.5).sin();
        let phi = s * s;

        let b_sum = self.b0 + self.b1 + self.b2;
        let a_sum = self.a0 + self.a1 + self.a2;

        let num = b_sum * b_sum
            - 4.0 * (self.b0 * self.b1 + 4.0 * self.b0 * self.b2 + self.b1 * self.b2) * phi
            + 16.0 * self.b0 * self.b2 * phi * phi;
        let den = a_sum * a_sum
            - 4.0 * (self.a0 * self.a1 + 4.0 * self.a0 * self.a2 + self.a1 * self.a2) * phi
            + 16.0 * self.a0 * self.a2 * phi * phi;

        if den <= 0.0 {
            return 0.0;
        }
        (num.max(0.0)) / den
    }

    pub fn gain_db(&self, w: f64) -> f64 {
        let m2 = self.magnitude_squared(w);
        if m2 <= 1e-20 {
            -200.0
        } else {
            10.0 * m2.log10()
        }
    }

    pub fn peaking(freq: f64, sample_rate: f64, gain_db: f64, q: f64) -> Self {
        let w0 = omega(freq, sample_rate);
        let a = 10f64.powf(gain_db / 40.0);
        let alpha = w0.sin() / (2.0 * q);
        let cos = w0.cos();

        Biquad {
            b0: 1.0 + alpha * a,
            b1: -2.0 * cos,
            b2: 1.0 - alpha * a,
            a0: 1.0 + alpha / a,
            a1: -2.0 * cos,
            a2: 1.0 - alpha / a,
        }
    }

    pub fn low_shelf(freq: f64, sample_rate: f64, gain_db: f64, q: f64) -> Self {
        let w0 = omega(freq, sample_rate);
        let a = 10f64.powf(gain_db / 40.0);
        let alpha = w0.sin() / (2.0 * q);
        let cos = w0.cos();
        let sqrt_a2_alpha = 2.0 * a.sqrt() * alpha;

        Biquad {
            b0: a * ((a + 1.0) - (a - 1.0) * cos + sqrt_a2_alpha),
            b1: 2.0 * a * ((a - 1.0) - (a + 1.0) * cos),
            b2: a * ((a + 1.0) - (a - 1.0) * cos - sqrt_a2_alpha),
            a0: (a + 1.0) + (a - 1.0) * cos + sqrt_a2_alpha,
            a1: -2.0 * ((a - 1.0) + (a + 1.0) * cos),
            a2: (a + 1.0) + (a - 1.0) * cos - sqrt_a2_alpha,
        }
    }

    pub fn high_shelf(freq: f64, sample_rate: f64, gain_db: f64, q: f64) -> Self {
        let w0 = omega(freq, sample_rate);
        let a = 10f64.powf(gain_db / 40.0);
        let alpha = w0.sin() / (2.0 * q);
        let cos = w0.cos();
        let sqrt_a2_alpha = 2.0 * a.sqrt() * alpha;

        Biquad {
            b0: a * ((a + 1.0) + (a - 1.0) * cos + sqrt_a2_alpha),
            b1: -2.0 * a * ((a - 1.0) + (a + 1.0) * cos),
            b2: a * ((a + 1.0) + (a - 1.0) * cos - sqrt_a2_alpha),
            a0: (a + 1.0) - (a - 1.0) * cos + sqrt_a2_alpha,
            a1: 2.0 * ((a - 1.0) - (a + 1.0) * cos),
            a2: (a + 1.0) - (a - 1.0) * cos - sqrt_a2_alpha,
        }
    }

    pub fn high_pass(freq: f64, sample_rate: f64, q: f64) -> Self {
        let w0 = omega(freq, sample_rate);
        let alpha = w0.sin() / (2.0 * q);
        let cos = w0.cos();

        Biquad {
            b0: (1.0 + cos) / 2.0,
            b1: -(1.0 + cos),
            b2: (1.0 + cos) / 2.0,
            a0: 1.0 + alpha,
            a1: -2.0 * cos,
            a2: 1.0 - alpha,
        }
    }

    pub fn low_pass(freq: f64, sample_rate: f64, q: f64) -> Self {
        let w0 = omega(freq, sample_rate);
        let alpha = w0.sin() / (2.0 * q);
        let cos = w0.cos();

        Biquad {
            b0: (1.0 - cos) / 2.0,
            b1: 1.0 - cos,
            b2: (1.0 - cos) / 2.0,
            a0: 1.0 + alpha,
            a1: -2.0 * cos,
            a2: 1.0 - alpha,
        }
    }

    pub fn notch(freq: f64, sample_rate: f64, q: f64) -> Self {
        let w0 = omega(freq, sample_rate);
        let alpha = w0.sin() / (2.0 * q);
        let cos = w0.cos();

        Biquad {
            b0: 1.0,
            b1: -2.0 * cos,
            b2: 1.0,
            a0: 1.0 + alpha,
            a1: -2.0 * cos,
            a2: 1.0 - alpha,
        }
    }

    pub fn band_pass(freq: f64, sample_rate: f64, q: f64) -> Self {
        let w0 = omega(freq, sample_rate);
        let alpha = w0.sin() / (2.0 * q);
        let cos = w0.cos();

        Biquad {
            b0: alpha,
            b1: 0.0,
            b2: -alpha,
            a0: 1.0 + alpha,
            a1: -2.0 * cos,
            a2: 1.0 - alpha,
        }
    }

    pub fn all_pass(freq: f64, sample_rate: f64, q: f64) -> Self {
        let w0 = omega(freq, sample_rate);
        let alpha = w0.sin() / (2.0 * q);
        let cos = w0.cos();

        Biquad {
            b0: 1.0 - alpha,
            b1: -2.0 * cos,
            b2: 1.0 + alpha,
            a0: 1.0 + alpha,
            a1: -2.0 * cos,
            a2: 1.0 - alpha,
        }
    }
}

fn omega(freq: f64, sample_rate: f64) -> f64 {
    let f = freq.min(sample_rate * 0.4999).max(1.0);
    2.0 * PI * f / sample_rate
}

pub fn butterworth_qs(order: usize) -> Vec<f64> {
    let mut order = order.max(2);
    if order % 2 != 0 {
        order += 1;
    }
    let sections = order / 2;
    (0..sections)
        .map(|k| 1.0 / (2.0 * ((2.0 * k as f64 + 1.0) * PI / (2.0 * order as f64)).cos()))
        .collect()
}
